### PR TITLE
fixes #5 - turn functional doc tests into unittests

### LINKED ISSUE #5: Remove dependency on zope.app.wsgi
The test-setup of the doctest uses zope.app.wsgi which pulls in dependencies we do not want (zope.app.publication, zope.session, zope.app.appsetup).

### PR DESCRIPTION
**Pull request by jensens**
The very basic functional tests are turned into plain unittests, which reduces the complex dependencies of test extra to zero. This also increases the depth/coverage of testing.

cc @pbauer

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -4,7 +4,8 @@ Changelog
 1.3.1 (unreleased)
 ------------------
 
-- Nothing changed yet.
+- Turn fucntional tests into better covering unit tests and also add more tests.
```
Comment (mgedmin) [inline CHANGES.rst:7]: Typo: fucntional -> functional.

I also think "better-covering" needs a hyphen, but I'm not a native English speaker and may be mistaken.

File: README.rst [hunk 2]
```diff
@@ -1 +1,15 @@
-Please refer to src/zope/globalrequest/README.rst.
+zope.globalrequest
+==================
+
+Introduction
+------------
+
+This package provides a global way to retrieve the currently active request object in a zope-based web framework.
```
Comment (mgedmin) [inline README.rst:7]: Maybe it's just me, but when I see "global" I have questions about thread safety.  Perhaps mention "the currently active request object of the current thread"?  Would that make the description too unwieldy?  Maybe put "(of the current thread)" in parentheses?

File: buildout.cfg [hunk 3]
```diff
@@ -1,13 +1,14 @@
 [buildout]
 develop = .
-parts = interpreter test
+parts = interpreter test
```
Comment (mgedmin) [inline buildout.cfg:3]: Please don't add trailing unnecessary whitespace.

File: buildout.cfg [hunk 4]
```diff
@@ -1,13 +1,14 @@
 [buildout]
 develop = .
-parts = interpreter test
+parts = interpreter test 
 
 [interpreter]
 recipe = zc.recipe.egg
 eggs = zope.globalrequest
-interpreter = python
+interpreter = zopepy
```
Comment (mgedmin) [inline buildout.cfg:8]: I'm curious about the motivation for this.

File: README.rst (hunk 2 above)
Comment (mgedmin) [reply README.rst:7]: (I've just realized that you merely moved the text from the other README.rst here.)

File: src/zope/globalrequest/__init__.py [hunk 5]
```diff
@@ -20,4 +20,4 @@ def setRequest(request):
 
 def clearRequest():
     """ clear the stored request object """
-    setRequest(None)
+    setRequest(marker)
```
Comment (mgedmin) [inline src/zope/globalrequest/__init__.py:23]: Is this a bugfix, or an API change, or ... ?

File: src/zope/globalrequest/tests.py [hunk 6]
```diff
@@ -1,37 +1,93 @@
...
-def zcml(source):
-    context = config.ConfigurationMachine()
-    xmlconfig.registerCommonDirectives(context)
-    xmlconfig.string(source, context)
+class TestGlobalrequest(unittest.TestCase):
 
+    def tearDown(self):
+        # reset the threading local context
+        from threading import local
+        import zope.globalrequest.local
```
Comment (mgedmin) [inline src/zope/globalrequest/tests.py:10]: Why not move the imports to the top of the module?

**Review by mgedmin — COMMENTED** [review]
Mostly LGTM.  Thanks!

File: src/zope/globalrequest/__init__.py (hunk 5 above)
Comment (jensens) [reply src/zope/globalrequest/__init__.py:23]: A little fix while testing. It is supposed to return the marker if the value is unset in order to make the ``factory`` work, but marker is different from None. I suppose nobody uses the factory feature in real life, but as I detected it I changed it to behave correctly.

File: buildout.cfg (hunk 4 above)
Comment (jensens) [reply buildout.cfg:8]: well, when you  create a ``virtualenv  .`` (as many developer do) then ``bin/python`` is already the virtualenv one and so buildout fails to override a used python file.

File: src/zope/globalrequest/tests.py (hunk 6 above)
Comment (jensens) [reply src/zope/globalrequest/tests.py:10]: Well, in tests I usually import test related code inside the test methods. This is probably a bit paranoid. OTOH it avoids strange and difficult to debug side effects with imports...

File: buildout.cfg (hunk 3 above)
Comment (jensens) [reply buildout.cfg:3]: I'll remove it on master (@hannosch merged it already).

File: src/zope/globalrequest/__init__.py (hunk 5 above)
Comment (mgedmin) [reply src/zope/globalrequest/__init__.py:23]: Probably worth mentioning the fix in the changelog then.

File: src/zope/globalrequest/tests.py (hunk 6 above)
Comment (mgedmin) [reply src/zope/globalrequest/tests.py:10]: So far I knew a couple of reasons to do this:

- A long time ago test module import errors were difficult to see -- they'd scroll by at the top of the test runner output, and be followed by the actual test run ending in "All 42 tests passed" or similar.  But test runners got fixed to report test module import errors properly, and we stopped importing inside test methods.

- I suppose when you import everything in the test method that makes it easier to move the test to a different module when refactoring.

I'd like to hear more about those "strange and difficult to debug side effects", if you've time.  Sounds interesting!

File: buildout.cfg (hunk 4 above)
Comment (mgedmin) [reply buildout.cfg:8]: Good reason!

(Many other zopefoundation projects use `interpreter = py` for the same reason.  I'd never seen `zopepy` before.)

(Unrelated: I stopped using `virtualenv .` when pip started cluttering my `.` with files like `pip-selfcheck.json`.  Now I `virtualenv .env` and then sometimes symlink individual scripts bin/foo ->  .env/bin/foo.  It's annoying to do that for every script, but symlinking the entire `bin` breaks the virtualenv.)

File: src/zope/globalrequest/tests.py (hunk 6 above)
Comment (jensens) [reply src/zope/globalrequest/tests.py:10]: Well, its been a while I had those and its probably not that interesting. Its primary circular imports or importing code early so possible errors are popping up on load time of the test and not on execution time, when you would like to expect them.

File: buildout.cfg (hunk 4 above)
Comment (jensens) [reply buildout.cfg:8]: If ``py`` is preferred we can change it. In Plone world I saw a lot ``zopepy``, so I dont care as long as it does not collide.

Comment (mgedmin) [reply buildout.cfg:8]: Nah, it's too minor to bother changing, I was just curious.
